hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -1341,8 +1341,10 @@ public HFileBlock readBlock(long dataBlockOffset, long onDiskBlockSize, final bo
         // Cache the block if necessary
         cacheConf.getBlockCache().ifPresent(cache -> {
           if (cacheBlock && cacheConf.shouldCacheBlockOnRead(category)) {
+            // Using the wait on cache during compaction and prefetching.
             cache.cacheBlock(cacheKey, cacheCompressed ? hfileBlock : unpacked,
-              cacheConf.isInMemory());
+              cacheConf.isInMemory(), isCompaction ||
+                (pread  && expectedBlockType == null && expectedDataBlockEncoding == null));
```

Comment:
Mind explain the logic behind these particular flags to determine when we are going to trigger the wait? And we don't need this when caching compressed blocks in line#1331 above?

Reply:
cacheOnly option is enough for detecting prefetch and isCompaction for compaction operations changed that in new commit.